src/libcore/ptr.rs

```diff
@@ -8,11 +8,59 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-// FIXME: talk about offset, copy_memory, copy_nonoverlapping_memory
-
-//! Raw, unsafe pointers, `*const T`, and `*mut T`.
+//! Manually manage memory through raw pointers.
 //!
 //! *[See also the pointer primitive types](../../std/primitive.pointer.html).*
+//!
+//! # Safety
+//!
+//! Many functions in this module take raw pointers as arguments and read from
+//! or write to them. For this to be safe, these pointers must be *valid*.
+//! Whether a pointer is valid depends on the operation it is used for
+//! (read or write), and the extent of the memory that is accessed (i.e.,
+//! how many bytes are read/written). Most functions use `*mut T` and `*const T`
+//! to access only a single value, in which case the documentation omits the size
+//! and implicitly assumes it to be `size_of::<T>()` bytes.
+//!
```

Comment:
Also, these are not volatile/atomic operations (except for `read_volatile`/`write_volatile`), so:

1. If another thread/DMA/whatever is performing conflicting accesses to the memory behind the pointer, that is hard UB.
2. Writing to "loopy memory" (i.e., 2 virtual addresses mapped to the same physical address) is also UB.

Reply:
`read_volatile`/`write_volatile` across thread boundaries _is_ UB.

Reply:
It feels more like "underdefined behavior" (from a "system" POV, it is hard to differentiate between a cache-coherent DMA and another thread).

Reply:
`volatile` has no bearing on multi-threading, so it is just as UB as "normal" accesses. The compiler is allowed to optimize very differently around these.

I will add a link to [atomic accesses](https://doc.rust-lang.org/beta/std/sync/atomic/index.html) for the multi-threading stuff.

Reply:
Better like this?